### ISSUE TITLE
Problem with math in PTLens

### ISSUE THREAD

**Issue by uchrisu** [lensfun/lensfun#1137]
Hello, I am writing a tool to find the distortion coefficients for a PTLens model nearly automatically. Therefore, I had a deeper look on the math and implementation in lensfun.

I think there is a problem in the implementation if it comes to the transformation of the parameters a,b, and c in "rescale_polynomial_coefficients".

The original PTLens transformation is given by
r_d = r_u ( a r_u^3 + b r_u^2 + c r_u + 1 − a − b − c )

While the stretching in the center is 
\partial r_d \partial r_u [at r_u = 0] = 1 - a - b - c =: d

Now it is said that the transformation of the parameters 
a -> a' = a/d^4
b -> b' = b/d^3
c -> c' = c/d^2
stretches r_u by 1/d. 

This is okay for the terms with higher order (2 and higher) of r_u, but especially the first order which is crucial in the center of the image does something completely different.

The new r_d is:
r_d = r_u ( a/d^4 r_u^3 + b/d^3 r_u^2 + c/d^2 r_u + 1 − a/d^4 − b/d^3 − c/d^2 )
An now \partial r_d \partial r_u [at r_u = 0] = 1 − a/d^4 − b/d^3 − c/d^2
which is NOT one as it should be.

It think the discrepancy here is not so large, since d is close to 1, but then d and this whole transformation of the parameters could be neglected at all. 
Did I get something wrong or is this really an issue in the math implementation of PTLens?

Maybe someone more familiar with the code has an answer? @bronger maybe?

Thanks for any response.

**Comment by uchrisu** [comment]
Okay, after reviewing more part of the code I realized what you have done. You simply changed the linear part to factor 1, not only changing, a,b,c. That's fine and math is happy again :)

Maybe I will update the comments in the code to clarify this.

**Comment by bronger** [comment]
All this trouble is because PTLens (coming from PanoTools/Hugin) is defined in a bad way.  “d” should be 1, so that ∂r_d/∂rᵤ = 1.  Now, it changes the focal length by 1-a-b-c, see https://bugs.launchpad.net/hugin/+bug/1307457.  Unfortnately, we need the correct focal length preserved for transformations like fisheye->rectilinear or perspective correction. This is the reason for the convoluted code.

It is tempting to re-calculate all parameters in the XML files and make the source code *much* easier to understand. On the other hand, people could not use Hugin anymore for getting distortion parameters, unless they re-scale the coefficients themselves.

**Comment by uchrisu** [comment]
Thanks for the explanation.

So, if I create profiles I should do it in the format with 1-a-b-c as linear stretching factor. I think that was originally done to keep the size in the smaller image size direction. 

So, looking at the distortion models ([https://lensfun.github.io/manual/latest/group__Lens.html#gaa505e04666a189274ba66316697e308e](url) only Poly3 and PTLens have this problem. Poly 5 is free from this problem, right?

**Comment by bronger** [comment]
Yes, only Poly3 and PTLens are affected.

As far as creating profiles is concerned, stick to Hugin’s way to do it and you are safe, i.e. its coordinate system and its model formula.

**Comment by uchrisu** [comment]
I wrote a little python script that does most of the job for me. It finds the straight lines itself with only some manual suggestions and writes all results to a file fitting the file format for your calibrate script. 
I tested the results. They are very very close to the hugin results and the resulting lines (also in darktable) are straight. 

Thanks for the help. It took a little time until I got through the whole (math-)code in lensfun.
